e2e/src/bin/historic_trades.rs

```diff
@@ -0,0 +1,304 @@
+    models::BatchId,
+};
+use e2e::cmd;
+use pricegraph::{Element, Pricegraph, TokenPair, U256};
+use std::{fs::File, io::Write, path::PathBuf};
+use structopt::StructOpt;
+
+const FULLY_FILLED_THRESHOLD: f64 = 0.95;
+const MIN_AMOUNT: u128 = pricegraph::MIN_AMOUNT as _;
+
+/// Common options for analyzing historic batch data.
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "historic_trades",
+    about = "Utility for comparing historic trades to price estimates.",
+    rename_all = "kebab"
+)]
+struct Options {
+    /// The events registry file store containing past exchange events.
+    #[structopt(long, env = "ORDERBOOK_FILE", parse(from_os_str))]
+    orderbook_file: PathBuf,
+
+    /// The output directory for the computed results.
+    #[structopt(long, env = "OUTPUT_DIR", default_value = "target", parse(from_os_str))]
+    output_dir: PathBuf,
+}
+
+fn main() -> Result<()> {
+    let options = Options::from_args();
+
+    let mut report = Report::new(File::create(options.output_dir.join("trades.csv"))?);
+    report.header()?;
+
+    cmd::for_each_batch(&options.orderbook_file, |history, batch| {
+        let auction_elements = history.auction_elements_for_batch(batch)?;
+        let settlement = history.settlement_for_batch(batch);
+
+        let new_orders = auction_elements
+            .iter()
+            .filter(|order| batch == order.valid.from)
+            .filter(|order| order.price.numerator != 0 && order.price.denominator > MIN_AMOUNT)
+            .filter(|order| order.balance > U256::from(MIN_AMOUNT));
+        for order in new_orders {
+            let settlement = settlement.as_ref();
+            let pricegraph = Pricegraph::new(
+                auction_elements
+                    .iter()
+                    .filter(|element| (element.user, element.id) != (order.user, order.id))
+                    .cloned(),
+            );
+
+            let meta = OrderMetadata::compute(settlement, order, &pricegraph);
+            let result = if meta.is_reasonably_priced_order() {
+                process_reasonable_order(settlement, order, &meta)
+            } else {
+                process_unreasonable_order(&meta)
+            };
+
+            report.record_result(batch, settlement, order, &meta, result)?;
+        }
+
+        Ok(())
+    })?;
+
+    report.summary()
+}
+
+fn process_reasonable_order(
+    settlement: Option<&Settlement>,
+    order: &Element,
+    meta: &OrderMetadata,
+) -> TradeResult {
+    let settlement = match settlement {
+        Some(value) => value,
+        None => return TradeResult::NoSolution,
+    };
+
+    match meta.fill_ratio() {
+        Some(trade_ratio) if trade_ratio > FULLY_FILLED_THRESHOLD => TradeResult::FullyMatched,
+        Some(_) => TradeResult::PartiallyMatched,
+        None => match find_settled_exchange_rate(settlement, order.pair) {
+            Some(xrate) if xrate > meta.limit_price => TradeResult::SkippedMatchableOrder,
```

Comment:
Do we have to adjust for fees here?

Reply:
Good point, exchange prices are before fees.